### ISSUE TITLE
Failed to open a MATLAB engine session on Windows

### ISSUE THREAD

**Issue by wangjie212** [JuliaInterop/MATLAB.jl#167]
Hello,

When calling MSession(), I get the following error on Windows:
ERROR: MEngineError("Failed to open a MATLAB engine session")
in MSession at MATLAB\cVrxc\src\engine.jl:23.

I'm using Matlab R2019b 64 bits. I can start Matlab normally.
I have set the env variable MATLAB_HOME and runned matlab -regserver in a Command Prompt as Administrator.

Anyone knows how to fix this?

**Comment by musm** [comment]
which julia version and platform:

**Comment by wangjie212** [comment]
Julia 1.2.0 Windows10

**Comment by ronubi** [comment]
Without explicitly calling `MSession`, can you just do 

    mat"ver"

to see if Julia will connect to the Matlab engine?

**Comment by wangjie212** [comment]
> Without explicitly calling `MSession`, can you just do
> 
> ```
> mat"ver"
> ```
> 
> to see if Julia will connect to the Matlab engine?

It reports the error:
MEngineError("failed to open MATLAB engine session")

**Comment by musm** [comment]
very odd.

**Comment by jacob-roth** [comment]
I had to do the following (which is manually/hardcoded for R2020b)
0. clone my version https://github.com/jacob-roth/MATLAB.jl.git
1. open REPL
2. from REPL, set ENV["MATLAB_ROOT"] = "C:\\Program Files\\MATLAB\\R2020b\\" (you will have to move it to this location if you dont want to edit my fork)
3. ] add "C:\\path\\to\\cloned\\MATLAB.jl"
4. ] build MATLAB
5. using MATLAB

**Comment by jacob-roth** [comment]
> > Without explicitly calling `MSession`, can you just do
> > ```
> > mat"ver"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to see if Julia will connect to the Matlab engine?
> 
> It reports the error: MEngineError("failed to open MATLAB engine session")

I now encounter this as well
```
julia> MSession()
┌ Warning: Confirm MATLAB is installed and discoverable.
└ @ MATLAB C:\Users\roth0674\.julia\packages\MATLAB\5RUjq\src\engine.jl:28
┌ Warning: Ensure `matlab -regserver` has been run in a Command Prompt as Administrator.
└ @ MATLAB C:\Users\roth0674\.julia\packages\MATLAB\5RUjq\src\engine.jl:30
ERROR: MATLAB.MEngineError("failed to open MATLAB engine session")
Stacktrace:
 [1] MSession(bufsize::Int64; flags::String)
   @ MATLAB C:\Users\roth0674\.julia\packages\MATLAB\5RUjq\src\engine.jl:34
 [2] MSession(bufsize::Int64)
   @ MATLAB C:\Users\roth0674\.julia\packages\MATLAB\5RUjq\src\engine.jl:22
 [3] assign_persistent_msession()
   @ MATLAB C:\Users\roth0674\.julia\packages\MATLAB\5RUjq\src\MATLAB.jl:73
 [4] MSession(bufsize::Int64; flags::String)
   @ MATLAB C:\Users\roth0674\.julia\packages\MATLAB\5RUjq\src\engine.jl:24
 [5] MSession
   @ C:\Users\roth0674\.julia\packages\MATLAB\5RUjq\src\engine.jl:22 [inlined]
 [6] MSession()
   @ MATLAB C:\Users\roth0674\.julia\packages\MATLAB\5RUjq\src\engine.jl:22
 [7] top-level scope
   @ REPL[18]:1
```
even after doing the steps (which seemed to previously work) above (https://github.com/JuliaInterop/MATLAB.jl/issues/167#issuecomment-1771766974)

**Comment by VoSiLk** [comment]
Hi, the above mentioned workaround does not work for me. I still get the same error message. Is there another workaround so I can run MATLAB 2020b from julia while in parallel MATLAB2022b installed on my machine?

**Comment by tqml** [comment]
Hi @VoSiLk !
What operating system and julia version are you using?

What you could try (if not already done):
1. Remove the MATLAB julia package
2. delete existing compiled for MATLAB in the `.julia/compiled/<version>/MATLAB` directory (the `.julia` dir should be in `%HOMEPATH%\.julia\` or `~/.julia`)
3. set the environment variable `MATLAB_ROOT` (or `MATLAB_HOME`) to the version you want to use. To check, the resolved path will be `$MATLAB_ROOT/bin/<os-arch>/matlab[.exe]` so check if this is the correct path (`<os-arch>` dependes on your system, e.g. `win64`)
4. install MATLAB.jl again
5. Check that the correct path was used. E.g. try the following julia code: `using MATLAB; println(MATLAB.matlab_libpath)`
   For me this prints `"/Applications/MATLAB_R2023b.app/bin/maca64"` (on macos)
6. If the version is correct, check if MATLAB.jl works

**Comment by tqml** [comment]
One more thing: If you're on Windows, there seems to be an extra step.
The correct MATLAB "server" version must be registered with windows. This should happen automatically at install but if multiple versions are installed this might be the issue

1. Open a new MATLAB window as administrator (with the version you want to use)
2. In MATLAB run `!matlab -regserver`
3. Close the MATLAB window

The version of MATLAB you opened, should now be the "default" version according to Windows.
So ensure, that this version and the one that MATLAB.jl uses, are the same.

Link to MATLAB docs:
https://de.mathworks.com/help/matlab/matlab_external/registering-matlab-software-as-a-com-server.html

**Comment by VoSiLk** [comment]
Thanks for your help. 

The extra step:
1. Open a new MATLAB window as administrator (with the version you want to use)
2. In MATLAB run !matlab -regserver
3. Close the MATLAB window

before 

1. Remove the MATLAB julia package
2. Delete existing compiled for MATLAB in the .julia/compiled/<version>/MATLAB directory (the .julia dir should be in %HOMEPATH%\.julia\ or ~/.julia)
3. set the environment variable MATLAB_ROOT (or MATLAB_HOME) to the version you want to use. To check, the resolved path will be $MATLAB_ROOT/bin/<os-arch>/matlab[.exe] so check if this is the correct path (<os-arch> dependes on your system, e.g. win64)
4. install MATLAB.jl again
5. Check that the correct path was used. E.g. try the following julia code: using MATLAB; println(MATLAB.matlab_libpath)
For me this prints "/Applications/MATLAB_R2023b.app/bin/maca64" (on macos)
6. If the version is correct, check if MATLAB.jl works

has done the trick :)

My OS is Win11.

**Comment by tqml** [comment]
Perfect! Thanks for trying (don't have a windows machine to test this) :) 

I will push an update to error message to make this more clear :)
